Add extra_properties to hive table properties

posulliv opened the pull request:
Fixes #954

sajjoseph commented:
This will allow users to specify different hive table properties that are not natively exposed in the current hive connector. 
Thanks @posulliv for the contribution.

sajjoseph commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:
```diff
@@ -962,6 +969,12 @@ public void createTable(ConnectorSession session, ConnectorTableMetadata tableMe
         tableProperties.put("totalSize", "-1");
 
+        // Extra properties
+        Map<String, String> extraProperties = HiveTableProperties.getExtraProperties(tableMetadata.getProperties());
```
nit: How about switching to static import HiveTableProperties.getExtraProperties

posulliv replied:
Yep, done!

posulliv commented:
@sajjoseph good catch on #1176. I had not covered the `SHOW CREATE TABLE` case. Updated to cover that based on #1176 and added this to the unit tests for `extra_properties` as well.

dain commented:
This was proposed a while back, and I don’ think it can be done that way.  The issue is to change a property you must do a read/update/write cycle that includes all existing properties, and that is very dangerous as you could drop (or restore) properties.

I'm not sure how you could do this with the existing table property system.  I would guess we need to extend this to support properties that look like `some_base_name.arbitrary_name`, where all of them have the same type.  So for this case you would end up with `extra_properties.auto_purge`...  Or maybe replace `List<PropertyMetadata<?>> getTableProperties()` on `Connector` with  something like:
```
    /**
     * @return the table properties for this connector
     * @deprecated use getTablePropertyProvider
     */
    default List<PropertyMetadata<?>> getTableProperties()
    {
        return emptyList();
    }

    /**
     * @return the table properties for this connector
     */
    default TablePropertyProvider getTablePropertyProvider()
    {
        return TablePropertyProvider.fixedTableProperties(getTableProperties());
    }

    interface TablePropertyProvider
    {
        Set<String> getKnownTablePropertyNames();

        Optional<PropertyMetadata<?>> getTableProperty(String name);

        static TablePropertyProvider fixedTableProperties(Collection<PropertyMetadata<?>> properties)
        {
            Set<String> knownTablePropertyNames = properties.stream()
                    .map(PropertyMetadata::getName)
                    .collect(Collectors.toUnmodifiableSet());
            Map<String, PropertyMetadata<?>> tableProperties = properties.stream()
                    .collect(Collectors.toUnmodifiableMap(PropertyMetadata::getName, identity()));
            return new TablePropertyProvider()
            {
                @Override
                public Set<String> getKnownTablePropertyNames()
                {
                    return knownTablePropertyNames;
                }

                @Override
                public Optional<PropertyMetadata<?>> getTableProperty(String name)
                {
                    return Optional.ofNullable(tableProperties.get(name));
                }
            };
        }
    }
```

Of course this would require changing TablePropertyManager also

posulliv commented:
thanks for having a look @dain. I mostly opened this PR to get support for the Hive `auto.purge` table property.

I also opened another PR that just adds that individual table property - https://github.com/trinodb/trino/pull/9457

Given this approach will require a lot more changes, do you think the PR that simply adds the `auto_purge` table property is the way we should go?

posulliv commented:
I closed #9457 and will try to apply the feedback to this PR. Thanks again for the comments.

posulliv commented:
@dain I see #9523 is now open. I will close this PR.

findinpath commented:
@posulliv this separation between Trino properties and "extra properties" may be rather misleading for the end users. How will a user now which property is a Trino property and which property is an arbitrary property?

findepi commented:
@findinpath the separation is actually important. Otherwise, how a user would know whether a property is "supposed to be honored by Trino", and which property is just a "dumb passthru" to the metastore?

billonahill commented:
We're also blocked on lack of Hive table property support when trying to specify JSON timestamp formats, as described in https://github.com/trinodb/trino/issues/4538. Do we have consensus on this approach?

jchoi614 commented:
FYI We added this patch on top of our Trino fork & deployed to production successfully, specifically for the purpose of specifying JSON timestamp formats. /cc @billonahill

billonahill commented:
Are there plans to merge this branch into the main Trino release? Or are we exploring other ways to support this? As @jchoi614 mentioned, we've applied this and are running in in production, but want to be sure we're in sync with future releases.

mosabua commented:
👋 @posulliv @findepi @findinpath and others - this PR has become inactive. Could you work together to move this forward?

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

colebow commented:
Closing this one out due to inactivity, but please reopen if you would like to pick this back up.

jchoi614 commented:
> Are there plans to merge this branch into the main Trino release? Or are we exploring other ways to support this? As @jchoi614 mentioned, we've applied this and are running in in production, but want to be sure we're in sync with future releases.

Following up on this - should we reopen this PR @billonahill ? I do not see the use of Hive extra properties in the current Trino release & this is a patch that has been working for us in production.

billonahill commented:
Yeah @jchoi614 we should. @colebow can we please re-open this one? We have been running this patch in our production environment for quite some time now and would like to have it part of Trino. Or if there is another way to provide a timestamp format for a given table we can use another another approach.

colebow commented:
> Yeah @jchoi614 we should. @colebow can we please re-open this one? We have been running this patch in our production environment for quite some time now and would like to have it part of Trino. Or if there is another way to provide a timestamp format for a given table we can use another another approach.

I'm not an authority here, but I'd make sure our PR author (@posulliv) is actually interested in picking this back up and getting it across the finish line when it hasn't been updated in over a year. If not, someone else who's interested should open a new PR with attribution in the commit message. Opening/closing a PR is superficial and only represents the status of the work - but if the author isn't interested in picking it up, it should likely remain closed.

mosabua commented:
@findinpath since you reopened.. does that mean @findepi and team are going to move this to merge proactively with @posulliv ?

billonahill commented:
@colebow @findepi @mosabua @posulliv we'd be glad to help get this across the finish line. @jchoi614 has a patch that we could put up in a new PR with attribution, or we can clean this one up so it will merge cleanly. Let us know which route you all prefer.

Praveen2112 commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:
```diff
@@ -8453,6 +8453,31 @@ public void testCreateAcidTableUnsupported()
         assertQueryFails("CREATE TABLE acid_unsupported WITH (transactional = true) AS SELECT 123 x", "FileHiveMetastore does not support ACID tables");
     }
 
+    @Test
```
Can we add additional test with altering table properties ? And what if we try to override one of the existing extra table properties ?

posulliv replied:
Do you mean a test with altering the table properties through hive? I can't write a test doing this with the hive connector in Trino since it doesn't support `ALTER TABLE <xyz> SET PROPERTIES`.

Praveen2112 commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:
```diff
@@ -1090,6 +1098,15 @@ else if (avroSchemaLiteral != null) {
+        Map<String, String> extraProperties = getExtraProperties(tableMetadata.getProperties());
+        if (extraProperties != null) {
+            tableProperties.put(EXTRA_PROPERTIES, mapJsonCodec(String.class, String.class).toJson(extraProperties));
+            for (Map.Entry<String, String> extraProperty : extraProperties.entrySet()) {
```
Can we have an additional check to ensure that - we don't override the existing property here ?

posulliv replied:
Would this be a test that alters the additional properties through hive?  I can't write a test doing this with the hive connector in Trino since it doesn't support `ALTER TABLE <xyz> SET PROPERTIES`.

posulliv replied:
I added a test for when duplicate properties are specified in `extra_properties` when creating a table.

Are you thinking that a user could add duplicate properties via an `ALTER TABLE` statement in hive? Is that possible?

Praveen2112 replied:
I was mentioning if the `extra_properties` could override any of the pre-defined property like file format or table_comment

posulliv replied:
Ah gotcha. No it is not possible for this to occur. I added an extra test for this scenario where we try to override `transactional` table property in `extra_properties`.

Praveen2112 commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:
```diff
@@ -8453,6 +8453,31 @@ public void testCreateAcidTableUnsupported()
+        assertQuery(
+                "SELECT * FROM \"test_extra_properties$properties\"",
+                "SELECT 'workaround for potential lack of HIVE-12730', 'false', 'true', '{\n  \"extra.property\" : \"true\"\n}', '0', '0', '" + queryId + "', '" + nodeVersion + "', '0', '0', 'false'");
+        MaterializedResult actualResult = computeActual("SHOW CREATE TABLE " + tableName);
+        assertTrue(getOnlyElement(actualResult.getOnlyColumnAsSet()).toString().contains("extra_properties = map_from_entries(ARRAY[ROW('extra.property', 'true')])"));
```
Instead of `contains` can we check the entire query String ?

huberty89 reviewed (approved):
Works fine for me

Praveen2112 commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java:
```diff
@@ -169,7 +173,16 @@ public HiveTableProperties(
+                new PropertyMetadata<>(
+                        EXTRA_PROPERTIES,
+                        "Extra table properties",
+                        typeManager.getType(TypeSignature.mapType(VARCHAR.getTypeSignature(), VARCHAR.getTypeSignature())),
```
Can we use `new MapType(VARCHAR, VARCHAR, typeManager.getTypeOperators())`

Praveen2112 commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveTableProperties.java:
```diff
@@ -169,7 +173,16 @@ public HiveTableProperties(
+                        "Extra table properties",
+                        typeManager.getType(TypeSignature.mapType(VARCHAR.getTypeSignature(), VARCHAR.getTypeSignature())),
+                        Map.class,
+                        null,
```
default value can be an emptyMap

Praveen2112 replied:
Sorry didn't see this one - https://github.com/trinodb/trino/pull/1176#issuecomment-662753140. This comment can be reverted.

Praveen2112 commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:
```diff
@@ -685,6 +688,11 @@ private ConnectorTableMetadata doGetTableMetadata(ConnectorSession session, Sche
         properties.putAll(partitionProjectionService.getPartitionProjectionTrinoTableProperties(table));
 
+        String extraProperties = table.getParameters().get(EXTRA_PROPERTIES);
+        if (extraProperties != null) {
+            properties.put(EXTRA_PROPERTIES, mapJsonCodec(String.class, String.class).fromJson(extraProperties));
```
What if hive/spark or a different system add some additional properties - is it possible for us to capture all of them ?

posulliv replied:
I'm not sure what you mean here?

Praveen2112 replied:
Say if we Hive/Spark adds some additional properties - it won't be visible here right ? Since this JsonString - captures only the extra-properties configured via Trino - Is this intentional or can we be a bit open about the properties added by other system ?

posulliv replied:
I see. Yes that is correct. It was intentional as we didn't have a use case for that. Do we want to add that as part of this PR?

Praveen2112 commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:
```diff
@@ -8453,6 +8453,84 @@ public void testCreateAcidTableUnsupported()
+                "SELECT * FROM \"test_multiple_extra_properties$properties\"",
+                "SELECT 'workaround for potential lack of HIVE-12730', 'false', 'one', 'two', '{\n  \"extra.property.one\" : \"one\",\n  \"extra.property.two\" : \"two\"\n}', '0', '0', '" + queryId + "', '" + nodeVersion + "', '0', '0', 'false'");
+        MaterializedResult actualResult = computeActual("SHOW CREATE TABLE " + tableName);
+        String expectedShowCreateTableSql = "CREATE TABLE hive.tpch.test_multiple_extra_properties (\n" +
+                "   c1 integer\n" +
+                ")\n" +
+                "WITH (\n" +
+                "   extra_properties = map_from_entries(ARRAY[ROW('extra.property.one', 'one'),ROW('extra.property.two', 'two')]),\n" +
+                "   format = 'ORC'\n" +
+                ")";
```
We can use text blocks for these String

posulliv replied:
@Praveen2112 I could not use text blocks for the expected output of `SHOW CREATE TABLE` as they would always trigger the `Whitespace character before closing parenthesis` check style warning. If I updated the string to resolve the check style warning then the test would fail because the string would not match.

Praveen2112 commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:
```diff
@@ -8453,6 +8453,84 @@ public void testCreateAcidTableUnsupported()
+    {
+        String tableName = format("%s.%s.test_duplicate_extra_properties", getSession().getCatalog().get(), getSession().getSchema().get());
+        @Language("SQL") String createTableSql = format("" +
+                        "CREATE TABLE %s (\n" +
+                        "  c1 integer\n" +
+                        ")\n" +
+                        "WITH (\n" +
+                        "  extra_properties = MAP(ARRAY['extra.property', 'extra.property'], Array['true', 'false'])\n," +
+                        "  format = 'ORC'\n" +
+                        ")",
+                tableName);
```
We can use text-blocks for this.

findinpath commented:
I was looking at https://trino.io/docs/current/sql/alter-table.html#set-properties documentation and as @findepi  mentioned in https://github.com/trinodb/trino/pull/9475#issuecomment-1061534055 Trino has its own table properties that it has to honour and that's why we can't really use this syntax for doing pass throughs for arbitrary properties.

The current PR solves partly this problem by adding the "extra_properties" workaround. However it is missing the fact that the users may add through Spark arbitrary properties which are not visible then in Trino.

Nevertheless, users need a way to pass through arbitrary properties so that they can land on the metastore.

An alternative idea would be to provide _system procedures_ to allow the user to perform such operations add/remove/list arbitrary table operations. This path doesn't involve any hacking on the `SET PROPERTIES` syntax.

findepi commented:
> The current PR solves partly this problem by adding the "extra_properties" workaround. However it is missing the fact that the users may add through Spark arbitrary properties which are not visible then in Trino.

`extra_properties` would expose all properties not understood by Trino, including those set after table creation


> An alternative idea would be to provide system procedures to allow the user to perform such operations add/remove/list arbitrary table operations. 

I think I am fine having that but I am concerned about users being able to shoot themselves in the foot.
But maybe that's OK.

Praveen2112 commented:
> An alternative idea would be to provide system procedures to allow the user to perform such operations add/remove/list arbitrary table operations.

But won't it affect the overall UX it would be like creating a table and then invoking the procedure to add/update custom property. 

> extra_properties would expose all properties not understood by Trino

But doesn't this allow us to migrate the table schema and related properties from one HMS to a different HMS. (One of the use case of `SHOW CREATE ..`) 

> Otherwise, how a user would know whether a property is "supposed to be honored by Trino", and which property is just a "dumb passthru" to the metastore?

I think we filter out the properties which are understood and honored by Trino and remaining properties could be exposed as `extra_properties`.

findinpath commented:
> But doesn't this allow us to migrate the table schema and related properties from one HMS to a different HMS. (One of the use case of `SHOW CREATE` ..)

@Praveen2112 even if we specify some "extra" properties in the `CREATE TABLE` statement,  the user will still be a bit confused because not all the properties will appear as expected in the `SHOW CREATE TABLE` statement.
See in https://github.com/trinodb/trino/pull/9475#discussion_r1119246009 `presto_query_id`, `presto_version`, etc properties which are created anyway on the fly.

We need to cover probably the need to set arbitrary properties when the table is being created. I don't think though that we need to showcase the "extra" properties in the `SHOW CREATE TABLE`.

posulliv commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:
```diff
@@ -8453,6 +8454,98 @@ public void testCreateAcidTableUnsupported()
+                ")\n" +
+                "WITH (\n" +
+                "   extra_properties = map_from_entries(ARRAY[ROW('extra.property', 'true')]),\n" +
+                "   format = 'ORC'\n" +
```
@findinpath are you asking why the additional hive table properties not supported by trino are not included in the extra properties in the `show create table` output?

If so, yes that was intentional. This PR was to only support extra properties that are specified explicitly by the user when the table is created through trino.

findinpath replied:
I'm thinking about the context where users do update from different engines the table properties.
Being limited to a number of "extra" properties defined up-front seems rather limiting to me.

Do note that a similar UX experience will need to be offered for Iceberg, Delta as well.

posulliv replied:
@findinpath got it. This PR was intentionally limited to the hive connector and wasn't intended to deal with other query engines.

Do you think that could be done as a follow up?

findinpath replied:
> Do you think that could be done as a follow up?

Definitely.
Let's first settle on how to do it in Hive and implementing the functionality in other connectors will eventually follow.

Praveen2112 replied:
> This PR was to only support extra properties that are specified explicitly by the user when the table is created through trino.

Can we extend this functionality in this PR ? One more additional concern here is that we add an additional properties `extra_properties` in Hive table  which can be overriden by other engines. We could skip some of the properties like `STATS_GENERATED_VIA_STATS_TASK`, `presto_query_id` and `presto_version`

Praveen2112 replied:
`CREATED_BY` and `TRINO_VIEW_FLAG`

posulliv commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:
```diff
@@ -8453,6 +8454,98 @@ public void testCreateAcidTableUnsupported()
+                .isInstanceOf(QueryFailedException.class)
+                .hasMessage("Multiple entries with same key: transactional=true and transactional=false");
+    }
+
```
@findinpath are you asking me to add a test that verifies the error when `null` property values are used? Or to do better error checking? Just want to be sure I understand the comment.

findinpath replied:
Let's first check whether setting `NULL` is valid. Check whether hive does accept setting properties `NULL` values.
In any case, the error thrown to the user should not be NPE.

posulliv replied:
hive will accept `null` as a string literal:

```
: jdbc:hive2://localhost:10000/default> create table t2(c1 int) tblproperties('extra.property'='null');
0: jdbc:hive2://localhost:10000/default>
```

The properties are then:

```
: jdbc:hive2://localhost:10000/default> show create table t2;
+----------------------------------------------------+
|                   createtab_stmt                   |
+----------------------------------------------------+
| CREATE TABLE `t2`(                                 |
|   `c1` int)                                        |
| ROW FORMAT SERDE                                   |
|   'org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe'  |
| STORED AS INPUTFORMAT                              |
|   'org.apache.hadoop.mapred.TextInputFormat'       |
| OUTPUTFORMAT                                       |
|   'org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat' |
| LOCATION                                           |
|   'hdfs://hadoop-master:9000/user/hive/warehouse/t2' |
| TBLPROPERTIES (                                    |
|   'bucketing_version'='2',                         |
|   'extra.property'='null',                         |
|   'transient_lastDdlTime'='1677618343')            |
+----------------------------------------------------+
14 rows selected (0.065 seconds)
0: jdbc:hive2://localhost:10000/default>
```

findinpath replied:
I just checked also on my own

```
0: jdbc:hive2://localhost:10000/default> create table t2(c1 int) tblproperties('extra.property'=null);
Error: Error while compiling statement: FAILED: ParseException line 1:55 mismatched input 'null' expecting StringLiteral near '=' in specifying key/value property (state=42000,code=40000)
```

Specifying `NULL` can't be used in Hive as well as a table property.

Praveen2112 replied:
In case of above `null` is represented as a `String`.

posulliv replied:
I thought @findinpath was testing with `null` as a string for an extra property which is what I tested but looks like he was not.

What changes do you want to see in the PR based on this discussion?

Praveen2112 replied:
We need to add a check if the property values are null, if null then we need to throw an `new TrinoException(INVALID_TABLE_PROPERTY...`

hashhar commented:
Obsoleted by https://github.com/trinodb/trino/pull/17172
